### PR TITLE
cephfs-mirror: synchronize directory snapshots

### PR DESCRIPTION
**Pull request by vshankar**
depends on PR #37721 

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/tools/cephfs_mirror/PeerReplayer.cc [hunk 1]
```diff
@@ -1049,6 +1067,33 @@ void PeerReplayer::run(SnapshotReplayerThread *replayer) {
...
+      f->open_object_section("last_synced_snap");
+      f->dump_unsigned("id", (*sync_stat.last_synced_snap).first);
+      f->dump_string("name", (*sync_stat.last_synced_snap).second);
+      if (sync_stat.last_sync_duration) {
+        f->dump_float("sync_duration", *sync_stat.last_sync_duration);
+      }
+      f->close_section();
+    }
+    f->dump_unsigned("snaps_synced", sync_stat.synced_snap_count);
+    f->dump_unsigned("snaps_deleted", sync_stat.deleted_snap_count);
```
Comment (mchangir) [inline src/tools/cephfs_mirror/PeerReplayer.cc:1093]: snaps renamed ?

**Review by mchangir — COMMENTED** [review]
a few nits
otherwise LGTM

File: src/client/Client.cc [hunk 2]
```diff
@@ -10985,7 +10985,8 @@ int Client::lazyio_synchronize(int fd, loff_t offset, size_t count)
 // =============================
 // snaps
 
-int Client::mksnap(const char *relpath, const char *name, const UserPerm& perm)
+int Client::mksnap(const char *relpath, const char *name, const UserPerm& perm,
+                   mode_t mode, const char *metadata_dict)
```
[the remark this answers is not in the document]
Comment (vshankar) [reply src/client/Client.cc:11040]: hmmm -- make sense. I'll include that change in PR #37721

File: src/client/Client.cc [hunk 3]
```diff
@@ -12726,13 +12729,28 @@ int Client::_mkdir(Inode *dir, const char *name, mode_t mode, const UserPerm& pe
...
+  if (is_snap_op) {
+    SnapPayload payload;
+    if (metadata_dict != nullptr) {
+      while (*metadata_dict != '\0') {
+        const char* key = metadata_dict;
+        metadata_dict += strlen(key) + 1;
+        const char* value = metadata_dict;
+        metadata_dict += strlen(value) + 1;
+        payload.metadata.emplace(key, value);
+      }
```
[the remark this answers is not in the document]
Comment (vshankar) [reply src/client/Client.cc:12797]: right!

File: src/tools/cephfs_mirror/PeerReplayer.cc (hunk 1 above)
Comment (vshankar) [reply src/tools/cephfs_mirror/PeerReplayer.cc:1093]: I thought about that, but failed to infer how would that be useful. Snaps synced and deleted are particularly useful, but would snaps renamed be of much use?

Comment (mchangir) [reply src/tools/cephfs_mirror/PeerReplayer.cc:1093]: I tend to think that stats are for audit purpose.
You may want to go through the rename stats to confirm that your actions have been accounted for ... no more and no less.

Comment (vshankar) [reply src/tools/cephfs_mirror/PeerReplayer.cc:1093]: Ummm... fair point. I'll include that too. Thx!

File: src/client/Client.cc (hunk 3 above)
[the remark this answers is not in the document]
Comment (vshankar) [reply src/client/Client.cc:12797]: FWIW, librados uses this way of passing metadata and it does not expect the caller to pass in the length.

**Comment by vshankar** [comment]
@batrick @ukernel please take a look.

**Comment by vshankar** [comment]
retest this please

**Comment by vshankar** [comment]
rebase!

**Comment by ajarr** [comment]
Few question while going through the doc/dev/cephfs-mirroring.rst,

- "cephfs_mirror_max_concurrent_failures_per_directory" this option name is unclear. Is it to configure max **concurrent** or **consecutive** retry failures? Can the option simply be, 'cephfs_mirror_max_retries_per_directory' or something like that?

- can the snapshot synchronization order be configured? or, does the mirror daemon pick snapshots for synchronization based on snap-id? 

- what is the current process to resolve `failed` synchronization that leaves incomplete data on a remote file system, and to resume synchronization of subsequent snapshots?

**Comment by vshankar** [comment]
> Few question while going through the doc/dev/cephfs-mirroring.rst,
> 
> * "cephfs_mirror_max_concurrent_failures_per_directory" this option name is unclear. Is it to configure max **concurrent** or **consecutive** retry failures? Can the option simply be, 'cephfs_mirror_max_retries_per_directory' or something like that?

The mirror daemon marks a directory as failed when synchronizing snapshots fail N consecutive times (where N is `cephfs_mirror_max_concurrent_failures_per_directory`). You are right -- it should probably be called `cephfs_mirror_max_consecutive_failures_per_directory`. or (as you suggested) `cephfs_mirror_max_retries_per_directory`.

> * can the snapshot synchronization order be configured? or, does the mirror daemon pick snapshots for synchronization based on snap-id?

Follows snapshot id.

> * what is the current process to resolve `failed` synchronization that leaves incomplete data on a remote file system, and to resume synchronization of subsequent snapshots?

Depends on what the failure is. If its a temporary failure such as a non-existent directory, the failure will be auto-resolved when the directory comes to existence. Other transient failures such as network and/or disk error should get resolved too (when the network is back up again, the disk is replaced, space is freed up, etc..). More permanent failures would involve the user to closely inspect the failures. The user can choose to remove the directory from synchronization till the issue is fixed and resolved (to avoid retries by the mirror daemon). We could provide some set of control mechanisms to "skip" a snapshot (for a directory) when it hits a failure. That would probably require additional metadata of some sort to be persisted. Right now, the mirror daemon ensures that a snapshot is synced successfully before moving to subsequent snapshots.

The incomplete data on the remote filesystem is left as-it-is which is ok since the remote filesystem should be treated read-only by the user (i.e., no active I/O should be in-progress).

**Comment by vshankar** [comment]
updated with Ramana's findings.

**Comment by vshankar** [comment]
one more thing @batrick : need review on the systemd unit files (top most commit).

File: doc/dev/cephfs-mirroring.rst [hunk 4]
```diff
@@ -22,6 +27,31 @@ This feature is tracked here: https://tracker.ceph.com/issues/47034
 .. note:: Synchronizing hardlinks is not supported -- hardlinked files get synchronized
           as separate files.
 
+Creating Users
+--------------
+
+Start by creating a user (on the primary/local cluster) for the mirror daemon. This user
+has restrictive capabilities on the MDS and the OSD::
+
+  $ ceph auth get-or-create client.mirror mon 'allow r' mds 'allow r' osd 'allow rw' mgr 'allow r'
```
Comment (batrick) [inline doc/dev/cephfs-mirroring.rst:36]: Why does it need `osd 'allow rw'`?

I would suggest just using `fs authorize` instead.

File: doc/dev/cephfs-mirroring.rst [hunk 5]
```diff
@@ -22,6 +27,31 @@ This feature is tracked here: https://tracker.ceph.com/issues/47034
...
+
+Create a user for each filesystem peer (on the secondary/remote cluster). This user needs
+to have full capabilities on the MDS (to take snapshots) and the OSDs::
+
+  $ ceph auth get-or-create client.mirror_remote mon 'allow r' mds 'allow rwps' osd 'allow rw' mgr 'allow r'
```
Comment (batrick) [inline doc/dev/cephfs-mirroring.rst:41]: ditto

File: doc/dev/cephfs-mirroring.rst [hunk 6]
```diff
@@ -22,6 +27,31 @@ This feature is tracked here: https://tracker.ceph.com/issues/47034
...
+
+This user should be used (as part of peer specification) when adding a peer.
+
+Starting Mirror Daemon
+----------------------
+
+Right now, mirror daemons need to be spawned manually. This will be replaced by providing
+`systemctl(1)` command to start/stop mirror daemons. For now, use::
```
Comment (batrick) [inline doc/dev/cephfs-mirroring.rst:49]: are the systemd units in this PR not usable currently?

File: doc/dev/cephfs-mirroring.rst [hunk 7]
```diff
@@ -242,7 +269,7 @@ mirror status. To check available commands for mirror status use::
   }
 
 Commands with `fs mirror status` prefix provide mirror status for mirror enabled
-filesystem. Note that `cephfs@360` is of format `filesystem-name@filesystem-id`.
+filesystems. Note that `cephfs@360` is of format `filesystem-name@filesystem-id`.
```
Comment (batrick) [inline doc/dev/cephfs-mirroring.rst:272]:
```suggestion
file systems. Note that `cephfs@360` is of format `filesystem-name@filesystem-id`.
```

**Review by batrick — CHANGES_REQUESTED** [review]
otherwise lgtm

File: doc/dev/cephfs-mirroring.rst (hunk 6 above)
Comment (vshankar) [reply doc/dev/cephfs-mirroring.rst:49]: They can be use (but needs review) -- the doc needs updation. I'll push an update.

File: doc/dev/cephfs-mirroring.rst (hunk 4 above)
Comment (vshankar) [reply doc/dev/cephfs-mirroring.rst:36]: mirror daemons need to create mirroring index object (`cephfs_mirror.<instance_id>`).

Comment (vshankar) [reply doc/dev/cephfs-mirroring.rst:36]: That's the reason to not use `fs authorize` interface -- we need write permission to the data pool (for creating the index object) but have the filesystem read only.

Comment (vshankar) [reply doc/dev/cephfs-mirroring.rst:36]: one more thing -- the user created here needs to be passed to the mirror daemon. `fs authorize` interface creates a user and grants permission to access a particular file system.

File: doc/dev/cephfs-mirroring.rst (hunk 5 above)
Comment (vshankar) [reply doc/dev/cephfs-mirroring.rst:41]: We can use `fs authorize` for creating users for the remote cluster/filesystem.

File: doc/dev/cephfs-mirroring.rst (hunk 4 above)
Comment (batrick) [reply doc/dev/cephfs-mirroring.rst:36]: `osd 'allow rw tag cephfs *=*'`

Comment (batrick) [reply doc/dev/cephfs-mirroring.rst:36]: `osd 'allow rw tag cephfs meta=*, allow r tag cephfs data=*'`

Comment (vshankar) [reply doc/dev/cephfs-mirroring.rst:36]: nice to know. thx!

Comment (vshankar) [reply doc/dev/cephfs-mirroring.rst:36]: > osd 'allow rw tag cephfs meta=*, allow r tag cephfs data=*'

works with `metadata=*`

**Review by batrick — APPROVED** [review]

**Comment by batrick** [comment]
https://pulpito.ceph.com/pdonnell-2021-01-12_22:03:59-fs-wip-pdonnell-testing-20210112.170550-distro-basic-smithi/

**Pull request merged**